### ISSUE TITLE
Reference Error When Compiling UnrealEngine.Runtime and UnrealEngine.MainDomain Solutions

### ISSUE THREAD

**Issue by Tanner555** [pixeltris/MonoUE-Standalone#1]
When I tried to compile the UnrealEngine.Runtime Solution, which is in my `MonoUE-Standalone\Managed\MonoBindings` folder, I got a reference error which causes the compile to fail. 
The Error is as follows:
```
C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\MSBuild\15.0\Bin\Microsoft.Common.CurrentVersion.targets(1179,5): error MSB3644: The reference assemblies for framework "MonoUE,Version=v1.0" were not found. To resolve this, install the SDK or Targeting Pack for this framework version or retarget your application to a version of the framework for which you have the SDK or Targeting Pack installed. Note that assemblies will be resolved from the Global Assembly Cache (GAC) and will be used in place of reference assemblies. Therefore your assembly may not be correctly targeted for the framework you intend.
```
![referenceerrorwhencompilingruntimesolution](https://user-images.githubusercontent.com/17560348/51228685-4fc41d80-190e-11e9-92c3-fdfaba0059da.PNG)

This error is caused by an issue in the SDK resolver and props files that are supposed to feed in all required references into the project. I installed the SDK resolver already, I can see it in my extensions and updates window in visual studio. I already had a conversation with pixeltris about this issue, and I believe he recommended manually adding these references to my project (I do not remember). 

The exact same error pops up when I try to compile the UnrealEngine.MainDomain solution as well.

**Comment by Tanner555** [comment]
To diagnose the issue, I've followed the instruction on one of the errors I encountered.
![designtimevsdebugging](https://user-images.githubusercontent.com/17560348/51231547-be0ddd80-1918-11e9-8690-9b35e09b854e.PNG)
After doing all this, I've reopened my solution, and looked for the indicated files in my local temp Appdata folder.
`UnrealEngine.Runtime_GetFrameworkPaths_3756ed69.designtime`
[UnrealEngine.Runtime_GetFrameworkPaths_3756ed69.designtime.log](https://github.com/pixeltris/MonoUE-Standalone/files/2762970/UnrealEngine.Runtime_GetFrameworkPaths_3756ed69.designtime.log)
`UnrealEngine.Runtime_Compile_37a3e1e8.designtime`
[Uploading UnrealEngine.Runtime_Compile_37a3e1e8.designtime.log…]()
`DTBB_d7ebc273-90f6-4e75-a047-86aa5664a751`
[DTBB_d7ebc273-90f6-4e75-a047-86aa5664a751.log](https://github.com/pixeltris/MonoUE-Standalone/files/2762972/DTBB_d7ebc273-90f6-4e75-a047-86aa5664a751.log)

**Comment by pixeltris** [comment]
Actually I'm not sure if that is related to the SDK resolver. I think it's saying it can't find `/Engine/Plugins/MonoUE-Standalone/ThirdParty/mono/fx/` or possibly `/Engine/Plugins/MonoUE-Standalone/ThirdParty/mono/fx/MonoUE/v1.0/`. It might be looking for referenced dlls in *both* of those locations. It does that lookup in [MonoUE.Core.props](https://github.com/pixeltris/MonoUE-Standalone/blob/fc49d66bf6fc0f1ea5f8de3e0660aee0a68d12b0/MSBuild/MonoUE.Core.props#L269).

**Comment by Tanner555** [comment]
Actually I looked at the location you mentioned `/Engine/Plugins/MonoUE-Standalone/ThirdParty/mono/fx/MonoUE/v1.0/` and I found all the dlls that these references should be looking up. I looked further into the log files, and discovered the problem may be an issue with msbuild. You can see that the ToolsVersion in the csproj file is 4.0, and for some reason the toolset cannot be found (I tried changing it to 15.0, doesn't work). In the Compile design time log file, you can see where the issue starts.
```
Building with tools version "15.0".
Project file contains ToolsVersion="4.0". This toolset may be unknown or missing, in which case you may be able to resolve this by installing the appropriate version of MSBuild, or the build may have been forced to a particular ToolsVersion for policy reasons. Treating the project as if it had ToolsVersion="15.0". For more information, please see http://go.microsoft.com/fwlink/?LinkId=293424.
Target "BeforeResolveReferences" in file "C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\MSBuild\15.0\Bin\Microsoft.Common.CurrentVersion.targets":
Done building target "BeforeResolveReferences" in project "UnrealEngine.Runtime.csproj".
Target "AssignProjectConfiguration" in file "C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\MSBuild\15.0\Bin\Microsoft.Common.CurrentVersion.targets":
  Using "AssignProjectConfiguration" task from assembly "Microsoft.Build.Tasks.Core, Version=15.1.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a".
  Task "AssignProjectConfiguration"
  Done executing task "AssignProjectConfiguration".
Done building target "AssignProjectConfiguration" in project "UnrealEngine.Runtime.csproj".
Target "AssignProjectConfiguration" skipped. Previously built successfully.
Target "_SplitProjectReferencesByFileExistence" in file "C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\MSBuild\15.0\Bin\Microsoft.Common.CurrentVersion.targets":
  Task "ResolveNonMSBuildProjectOutput" skipped, due to false condition; ('$(BuildingInsideVisualStudio)'=='true' and '@(ProjectReferenceWithConfiguration)'!='') was evaluated as ('true'=='true' and ''!='').
Done building target "_SplitProjectReferencesByFileExistence" in project "UnrealEngine.Runtime.csproj".
Target "_GetProjectReferenceTargetFrameworkProperties" in file "C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\MSBuild\15.0\Bin\Microsoft.Common.CurrentVersion.targets":
  Using "MSBuild" task from assembly "Microsoft.Build.Tasks.Core, Version=15.1.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a".
  Task "MSBuild"
  Done executing task "MSBuild".
  Task "GetReferenceNearestTargetFrameworkTask" skipped, due to false condition; ('@(_ProjectReferenceTargetFrameworkPossibilities->Count())' != '0' and '$(ReferringTargetFrameworkForProjectReferences)' != '') was evaluated as ('0' != '0' and 'MonoUE,Version=v1.0' != '').
Done building target "_GetProjectReferenceTargetFrameworkProperties" in project "UnrealEngine.Runtime.csproj".
Target "PrepareProjectReferences" in file "C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\MSBuild\15.0\Bin\Microsoft.Common.CurrentVersion.targets":
Done building target "PrepareProjectReferences" in project "UnrealEngine.Runtime.csproj".
Target "ResolveProjectReferences" in file "C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\MSBuild\15.0\Bin\Microsoft.Common.CurrentVersion.targets":
  Task "MSBuild" skipped, due to false condition; ('%(_MSBuildProjectReferenceExistent.BuildReference)' == 'true' and '@(ProjectReferenceWithConfiguration)' != '' and ('$(BuildingInsideVisualStudio)' == 'true' or '$(BuildProjectReferences)' != 'true') and '$(VisualStudioVersion)' != '10.0' and '@(_MSBuildProjectReferenceExistent)' != '') was evaluated as ('' == 'true' and '' != '' and ('true' == 'true' or 'true' != 'true') and '15.0' != '10.0' and '' != '').
  Task "MSBuild" skipped, due to false condition; ('%(_MSBuildProjectReferenceExistent.BuildReference)' == 'true' and '@(ProjectReferenceWithConfiguration)' != '' and ('$(BuildingInsideVisualStudio)' == 'true' or '$(BuildProjectReferences)' != 'true') and '$(VisualStudioVersion)' == '10.0' and '@(_MSBuildProjectReferenceExistent)' != '') was evaluated as ('' == 'true' and '' != '' and ('true' == 'true' or 'true' != 'true') and '15.0' == '10.0' and '' != '').
  Task "MSBuild" skipped, due to false condition; ('%(_MSBuildProjectReferenceExistent.BuildReference)' == 'true' and '@(ProjectReferenceWithConfiguration)' != '' and '$(BuildingInsideVisualStudio)' != 'true' and '$(BuildProjectReferences)' == 'true' and '@(_MSBuildProjectReferenceExistent)' != '') was evaluated as ('' == 'true' and '' != '' and 'true' != 'true' and 'true' == 'true' and '' != '').
  Task "MSBuild" skipped, due to false condition; ('%(_MSBuildProjectReferenceExistent.BuildReference)' == 'true' and '@(ProjectReferenceWithConfiguration)' != '' and '$(BuildingProject)' == 'true' and '@(_MSBuildProjectReferenceExistent)' != '') was evaluated as ('' == 'true' and '' != '' and 'false' == 'true' and '' != '').
  Task "Warning" skipped, due to false condition; ('@(ProjectReferenceWithConfiguration)' != '' and '@(_MSBuildProjectReferenceNonexistent)' != '') was evaluated as ('' != '' and '' != '').
Done building target "ResolveProjectReferences" in project "UnrealEngine.Runtime.csproj".
Target "FindInvalidProjectReferences" skipped, due to false condition; ('$(FindInvalidProjectReferences)' == 'true') was evaluated as ('' == 'true').
Target "ResolveNativeReferences" skipped, due to false condition; ('@(NativeReference)'!='') was evaluated as (''!='').
Target "GenerateMonoUE4Helpers" skipped, due to false condition; ('$(_MonoUEIsBindingAssembly)'=='False' And '$(UE4GameLocation)' != '') was evaluated as ('True'=='False' And '' != '').
Target "ResolveProjectReferences" skipped. Previously built successfully.
Target "FindInvalidProjectReferences" skipped, due to false condition; ('$(FindInvalidProjectReferences)' == 'true') was evaluated as ('' == 'true').
Target "GetFrameworkPaths" in file "C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\MSBuild\15.0\Bin\Microsoft.Common.CurrentVersion.targets":
Done building target "GetFrameworkPaths" in project "UnrealEngine.Runtime.csproj".
Target "GetFrameworkPaths" skipped. Previously built successfully.
Target "GetReferenceAssemblyPaths" in file "C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\MSBuild\15.0\Bin\Microsoft.Common.CurrentVersion.targets":
  Using "GetReferenceAssemblyPaths" task from assembly "Microsoft.Build.Tasks.Core, Version=15.1.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a".
  Task "GetReferenceAssemblyPaths"
    C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\MSBuild\15.0\Bin\Microsoft.Common.CurrentVersion.targets(1179,5): error MSB3644: The reference assemblies for framework "MonoUE,Version=v1.0" were not found. To resolve this, install the SDK or Targeting Pack for this framework version or retarget your application to a version of the framework for which you have the SDK or Targeting Pack installed. Note that assemblies will be resolved from the Global Assembly Cache (GAC) and will be used in place of reference assemblies. Therefore your assembly may not be correctly targeted for the framework you intend.
  Done executing task "GetReferenceAssemblyPaths" -- FAILED.
Done building target "GetReferenceAssemblyPaths" in project "UnrealEngine.Runtime.csproj" -- FAILED.

Done building project "UnrealEngine.Runtime.csproj" -- FAILED.
```
In the GetFrameworksPaths designtime log file, there's a similar issue near the end.
```
Building with tools version "15.0".
Project file contains ToolsVersion="4.0". This toolset may be unknown or missing, in which case you may be able to resolve this by installing the appropriate version of MSBuild, or the build may have been forced to a particular ToolsVersion for policy reasons. Treating the project as if it had ToolsVersion="15.0". For more information, please see http://go.microsoft.com/fwlink/?LinkId=293424.
Target "GetFrameworkPaths" in file "C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\MSBuild\15.0\Bin\Microsoft.Common.CurrentVersion.targets":
Done building target "GetFrameworkPaths" in project "UnrealEngine.Runtime.csproj".
```

There's a command somewhere in the props files, called "GetReferenceAssemblyPaths", which fails on execution.

**Comment by pixeltris** [comment]
I don't think the `ToolsVersion` is the cause of the problem. I'm fairly sure it just isn't able to find the dlls in the specified paths for some reason. `GetReferenceAssemblyPaths` is an MSBuild thing which I'm fairly sure is doing the dll lookups from the mentioned paths (`/Engine/Plugins/MonoUE-Standalone/ThirdParty/mono/fx/` or `/Engine/Plugins/MonoUE-Standalone/ThirdParty/mono/fx/MonoUE/v1.0/`).

You could try deleting that PropertyGroup which sets up the custom TargetFrameworkIdentifier stuff which might let the UnrealEngine.Runtime project compile but you will fail to compile game projects.

**Comment by Tanner555** [comment]
You are completely right. In the MonoUE.Core.props file, I tried commenting out the Property Group on the bottom here.
```
 <!-- set up the target framework -->
  <!-- <PropertyGroup>
    <TargetFrameworkIdentifier>MonoUE</TargetFrameworkIdentifier>
    <TargetFrameworkVersion>v1.0</TargetFrameworkVersion> -->
    <!-- <TargetFrameworkRootPath>$([System.IO.Path]::GetFullPath('$(MonoUEPluginDir)ThirdParty\mono\fx'))</TargetFrameworkRootPath> -->
    <!-- <FrameworkPathOverride>$(TargetFrameworkRootPath)\$(TargetFrameworkIdentifier)\$(TargetFrameworkVersion)</FrameworkPathOverride> -->
  <!-- </PropertyGroup> -->
```
I deleted my .vs and obj folders, then reopened the runtime solution. The initial reference resolving warning went completely away, and when I go to references, the dlls now point to the ones inside the .NET Framework reference assemblies path. When I try to build the solution, I get a few errors, none of which relate to the references not being found.
```
------ Build started: Project: UnrealEngine.Runtime, Configuration: Debug Any CPU ------
C:\Program Files\Epic Games\UE_4.20\Engine\Plugins\MonoUE-Standalone\Managed\MonoBindings\UnrealObject.cs(103,39,103,57): error CS0117: 'MethodImplOptions' does not contain a definition for 'AggressiveInlining'
C:\Program Files\Epic Games\UE_4.20\Engine\Plugins\MonoUE-Standalone\Managed\MonoBindings\FixedSizeArray.cs(19,16,19,43): error CS0308: The non-generic type 'WeakReference' cannot be used with type arguments
C:\Program Files\Epic Games\UE_4.20\Engine\Plugins\MonoUE-Standalone\Managed\MonoBindings\UnrealArray.cs(201,63,201,79): error CS0246: The type or namespace name 'IReadOnlyList<>' could not be found (are you missing a using directive or an assembly reference?)
C:\Program Files\Epic Games\UE_4.20\Engine\Plugins\MonoUE-Standalone\Managed\MonoBindings\UnrealArray.cs(397,87,397,103): error CS0246: The type or namespace name 'IReadOnlyList<>' could not be found (are you missing a using directive or an assembly reference?)
========== Build: 0 succeeded, 1 failed, 0 up-to-date, 0 skipped ==========
```
I initially thought the issue is probably caused from the compiler not being able to find the appropriate folder. I tried removing two lines responsible for finding the folder and assigning the FrameworkOverridePath.
```
  <PropertyGroup>
    <TargetFrameworkIdentifier>MonoUE</TargetFrameworkIdentifier>
    <TargetFrameworkVersion>v1.0</TargetFrameworkVersion>
    <!-- <TargetFrameworkRootPath>$([System.IO.Path]::GetFullPath('$(MonoUEPluginDir)ThirdParty\mono\fx'))</TargetFrameworkRootPath> -->
    <!-- <FrameworkPathOverride>$(TargetFrameworkRootPath)\$(TargetFrameworkIdentifier)\$(TargetFrameworkVersion)</FrameworkPathOverride> -->
  </PropertyGroup>
```
After deleting the .vs and obj folders, and reopening the runtime solution, I got the same reference warnings and errors as before. Perhaps the props file only applies the override if the identifier or framework version is set, I'm not sure. The folder mentioned does exist on my harddrive, exactly where it's indicated, so I do not know how it wouldn't be able to find it.

**Comment by pixeltris** [comment]
Those properties are mostly for MSBuild itself which are used for overriding the path where the framework dlls are. UnrealEngine.Runtime.csproj doesn't have a fallback `TargetFrameworkVersion` value provided so when you remove the override properties it probably ends up using .NET 4.0 or lower which doesn't contain things such as `MethodImplOptions.AggressiveInlining` as seen in your errors.

Again you could hack it by adding a `TargetFrameworkVersion` property in the csproj but this will only get you so far. It's best to try and find out why it can't resolve the dlls from the framework override path. Unfortunately the errors for WHY it can't resolve the dlls are basically non-existent.

It should be noted that I didn't change any of this stuff from the original MonoUE. I only modified the C++ and a small bit of the C# SDK resolver. If you have a regular MonoUE install you could compare folder paths and see if there is anything different which may be causing the issue.

**Comment by Tanner555** [comment]
WOW! 😆 
In Mono-UE, there's one xml file that was accidentally ignored by the gitignore, called "FrameworkList". It's inside of the `MonoUE-Standalone\ThirdParty\mono\fx\MonoUE\v1.0\RedistList` folder. This file exist in the MonoUE repo, but not inside of this one. The file contains this
```
<?xml version="1.0" encoding="utf-8"?>
<FileList Name="Mono for Unreal Engine 1.0" />
```
I took the file and put it inside my standalone fork, and when I opened the runtime solution (without any changes), I was able to compile the solution with no issues. I cannot believe I spent all of this time debugging visual studio, just to find out I was missing one XML file from MonoUE. We need to somehow not ignore this file. I'll keep this issue open until I include it in my PR and do further testing. Thanks for helping me out Pixel. 😄
